components/sensor/sht3xd/sht3xd.go

```diff
@@ -0,0 +1,182 @@
+// Package sht3xd implements a sht3x-d sensor for temperature and humidity
+package sht3xd
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/edaniels/golog"
+	"go.uber.org/multierr"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/components/board"
+	"go.viam.com/rdk/components/generic"
+	"go.viam.com/rdk/components/sensor"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/resource"
+	rdkutils "go.viam.com/rdk/utils"
+)
+
+var modelname = resource.NewDefaultModel("sht3xd")
+
+const (
+	defaultI2Caddr = 0x44
+	// Addresses of sht3xd registers.
+	sht3xdCOMMANDSOFTRESET1 = 0x30
+	sht3xdCOMMANDSOFTRESET2 = 0xA2
+	sht3xdCOMMANDPOLLINGH1  = 0x24
+	sht3xdCOMMANDPOLLINGH2  = 0x00
+)
+
+// AttrConfig is used for converting config attributes.
+type AttrConfig struct {
+	Board   string `json:"board"`
+	I2CBus  string `json:"i2c_bus"`
+	I2cAddr int    `json:"i2c_addr,omitempty"`
+}
+
+// Validate ensures all parts of the config are valid.
+func (config *AttrConfig) Validate(path string) ([]string, error) {
+	var deps []string
+	if len(config.Board) == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "board")
+	}
+	deps = append(deps, config.Board)
+	if len(config.I2CBus) == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "i2c bus")
+	}
+	return deps, nil
+}
+
+func init() {
+	registry.RegisterComponent(
+		sensor.Subtype,
+		modelname,
+		registry.Component{Constructor: func(
+			ctx context.Context,
+			deps registry.Dependencies,
+			config config.Component,
+			logger golog.Logger,
+		) (interface{}, error) {
+			attr, ok := config.ConvertedAttributes.(*AttrConfig)
+			if !ok {
+				return nil, rdkutils.NewUnexpectedTypeError(AttrConfig{}, config.ConvertedAttributes)
+			}
+			return newSensor(ctx, deps, config.Name, attr, logger)
+		}})
+
+	config.RegisterComponentAttributeMapConverter(sensor.Subtype, modelname,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf AttrConfig
+			return config.TransformAttributeMapToStruct(&conf, attributes)
+		}, &AttrConfig{})
+}
+
+func newSensor(
+	ctx context.Context,
+	deps registry.Dependencies,
+	name string,
+	attr *AttrConfig,
+	logger golog.Logger,
+) (sensor.Sensor, error) {
+	b, err := board.FromDependencies(deps, attr.Board)
+	if err != nil {
+		return nil, fmt.Errorf("sht3xd init: failed to find board: %w", err)
+	}
+	localB, ok := b.(board.LocalBoard)
+	if !ok {
+		return nil, fmt.Errorf("board %s is not local", attr.Board)
+	}
+	i2cbus, ok := localB.I2CByName(attr.I2CBus)
+	if !ok {
+		return nil, fmt.Errorf("sht3xd init: failed to find i2c bus %s", attr.I2CBus)
+	}
+	addr := attr.I2cAddr
+	if addr == 0 {
+		addr = defaultI2Caddr
+		logger.Warn("using i2c address : 0x44")
+	}
+
+	s := &sht3xd{
+		name:   name,
+		logger: logger,
+		bus:    i2cbus,
+		addr:   byte(addr),
```

Comment:
[minor] the datasheet and examples I looked at included information about setting a polling time, and a time constant for I guess reading the sensor. Would it be best to run that as a background goroutine here rather than open up the bus and close it each time we read?

The poll time examples I've seen from a cursory google search are around 60s, but can be configured. @agreenb, any poll time concerns on data's side?

Reply:
We've set frequencies of up to 16kHz on arm and gantry, but this was a while back and will be benchmarking again soon. Each reading is kicked off in its own goroutine so not blocked by a separate sensor reading — along with each reading, we also store the time the reading was requested and when it was received. It's mostly constrained by the hardware.

Reply:
In high repeatability/accuracy mode, this sensor can require up to 15ms for the operation, so that would max out at 66hz.

I cannot think of any use case where someone would be using this specific sensor, and would want temperature data at 66hz; I'm not convinced that the thermal characteristics of the chip would be such that such a high temperature resolution would be meaningful.

If you'd prefer that this be updated to use a background goroutine (similar to what the ultrasonic sensor uses?) I can do that, but as written it's similar to the bme280 and ds18b20 sensors we have.

Reply:
Let's go ahead and leave it as is for now. If we need to change it into a background goroutine (movement or ultrasonic sensors) and a polling time coming in from data or its own config, we'll do that as it becomes necessary.